### PR TITLE
docs: Add outline of unit testing recommendations

### PR DESCRIPTION
**Pull request by lundybernard**
This PR contains an outline of a proposed Development Guide / Principles / Unit Testing Recommendations section.

I am opening this PR now to allow discussion of the outline, and key-points, before fleshing out the complete page.

----
📚 Documentation preview 📚: https://scientific-python-cookie--619.org.readthedocs.build/

### CODE REVIEW

**Comment by henryiii** [comment]
Over, looks good, just one overall issue. I do not think we should push anyone toward unittest over pytest for unit tests. The reasons:

* pytest is the most popular framework. unittest is mostly there so CPython can test itself, and has a few specific uses. It's particularly bad for scientific work.
* Keeping the tests as easy to write as possible means more tests are written. unittest has a ton of boiler plate.
* Fixtures, parametrization, etc. are all _much_ better in pytest than in any xtest style framework.
* Installing a dependency to run unit tests is fine. Developers run unit tests.

That last point is important: unit tests are for developers to develop the code. End users do not need to run unit tests. Once the units work correctly, then you don't need to verify that again.

However, I think we should instead emphasize a new category of tests, the one we mentioned (I don't remember your term, "smoke test" is what I thought of). That's a separate category from unit tests, and for that one you don't want dependencies; it's a great place to use unittest. So:

* Integration tests: runs the entire app. Smaller projects often put them in `/tests`, larger projects might place them in a separate repository. Uses pytest or a command line runner. Can be slow.
* Unit tests: verifies the various parts work individually. Usually in `/tests`. Uses pytest. Should be reasonable to run often by the developer.
* Smoke test. Should be in the package, and runnable from the package. Uses unittest. Most common in compiled packages. Should verify the build works, and any architecture specific behavior works. Should be very fast.

So I'd basically recommend moving the unittest parts to a new section about smoke tests; the actual content is fine, just I think it needs reordering.

(And replace "smoke" with whatever you mentioned today, I've just forgotten what it was. Validity? Verification? ...)

File: docs/pages/principles/testing.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+## Unit Tests
```
Comment (henryiii) [inline docs/pages/principles/testing.md:1]: As mentioned above, I'd make this three sections. One short one about Integration testing. I think we could discuss ideas of cross-package integration testing there, like how to test that downstream packages still work, etc. Half of the below would go under "Unit tests", and the other half under the new "smoke" tests.

File: docs/pages/principles/testing.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+- Unit tests ensure that the code, as written, is correct, and executes
+  properly.
+- They communicate the intention of the creator of the code, how the code is
+  expected to behave, in its expected use-case.
+
+- Writing unit tests can reveal weakensses in our implementations, and lead us
+  to better design decisions:
+  - If the test requires excessive setup, the unit may be dependent on too many
+    external variables.
```
Comment (henryiii) [inline docs/pages/principles/testing.md:13]:
```suggestion
  - If the test requires excessive setup, the unit may be dependent on too many
    external variables. You might benefit from refactoring the code.
```

I think it would help here to make sure the reader knows the problme is in the code ("unit" here), not the test. That's clear in the last point, but I'd make sure it's obvious here too. Maybe this could be repeated every bullet point to drive it in.

File: docs/pages/principles/testing.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    understand and maintain. Refactoring code to make it easier to test often
+    leads us to write better code overall.
+
+### Guidelines for unit testing:
+
+- Unit tests live alongside the code they test, in a /tests folder.
+  - It looks like `unittest discover` cannot find TestCase classes within source
+    files.
+
+- Test files should be named `test_{{file under test}}.py`, so that stdlib
```
Comment (henryiii) [inline docs/pages/principles/testing.md:26]: I'd drop the mention of unittest. The naming is to help the user know what is being tested. Maybe "Test frameworks often require "test" in the file name" - you could keep this framework agnostic, and instead discuss frameworks in a separate section.

File: docs/pages/principles/testing.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  unittest can find them easily.
+
+- test\_.py files should match your source files (file-under-test) one-to-one,
+  and contain only tests for code in the file-file-under test.
```
Comment (henryiii) [inline docs/pages/principles/testing.md:30]: file-file-under? Also, `test_*.py` looks better, I think. I've seen people try to name a test file `test.py`. Doesn't end well. :)

File: docs/pages/principles/testing.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+- Consider using the stdlib `unittest.TestCase` and other stdlib tools instead
+  of pytest.
+  - Allows running unit tests for diagnostics in production environments,
+    without installing additional packages.
+
```
Comment (henryiii) [inline docs/pages/principles/testing.md:36]: This is exactly true under the smoke test section. Not under unit tests!

File: docs/pages/principles/testing.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+#### Importing in test files:
+
+- Use relative imports, from the file under test:
+
+  ```
+  from ..file_under_test import MyClass
+  ```
+
+  - This ensures that the test file always imports from its source file, and
+    does not accidentally import from an installed module.
```
Comment (henryiii) [inline docs/pages/principles/testing.md:52]: _NEVER DO THIS_! Only, only test the installed modules. Users install for pip/conda, etc. They do not download your repository and run files from there unless they are you. That's why we require the `/src` structure, to ensure no one every accidentally skips the packaging install when developing, so everything behaves just like what users see. 

(However, this is fine for the "smoke test" idea, since those actually live inside the package)

File: docs/pages/principles/testing.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+#### Running unit tests:
+
+- Pytest is great for running tests in your development environments!
+  `pytest {{path/to/source}}`
+- stdlib's unittest can be used in environments where pytest is not available:
+  - To use unittest to run tests in your source folder, from your package root,
+    use
+    `python -m unittest discover --start-folder {{source folder}} --top-level-directory .`
+  - To use unittest to run tests from an installed package (outside of your
+    source repository), use `python -m unittest discover -s {{module.name}}`
```
Comment (henryiii) [inline docs/pages/principles/testing.md:69]: I'd split this - if you use pytest to write the tests, you need to use it to run the tests, so not needed for the unit test section. But we do need a section to explain how to nicely set up the smoke tests to run. I can try to set this up somewhere and report back later on what seems to work well. `numpy.test()` might be a good example of what I'm thinking for smoke tests (I know that's really the whole unit test suite, but I think there's no need to put the entire unit test suite inside every wheel, just the validity tests).

File: docs/pages/principles/testing.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+- When the unit you are testing touches any external unit (usually something you
+  imported, or another unit that has its own tests), the external unit should be
+  Patched, replacing it with a Mock for the durration of the test.
+
+- Verify that the external unit is called with the expected input
+- Verify that any value returned from the external unit is utilized as expected.
+
+  ```
+  @patch(f'{SRC}.otherfunction', autospec=true)
```
Comment (henryiii) [inline docs/pages/principles/testing.md:81]: I'd show how to use pytest mocking here instead of unittest, or maybe better yet stay framework agnostic and just convey the ideas; the guide page on pytest already covers a bit of mocking and we can put more there.

**Comment by henryiii** [comment]
The only other thing I'd add is maybe a mention somewhere that this is guidelines for good design, but even poorly structured tests are better than no tests. I don't want to discourage someone from writing tests, but instead give them ways to improve their test design if they want to, and help as users scale up to larger projects.

File: docs/pages/principles/testing.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+- Consider using the stdlib `unittest.TestCase` and other stdlib tools instead
+  of pytest.
+  - Allows running unit tests for diagnostics in production environments,
+    without installing additional packages.
+
+- Keep it simple! If a test-case requires extra setup and external tools, It may
+  be more appropriate as an external test, instead of in the unit tests
+
+- Test single units of code! A single Function, or a single attribute or method
+  on a class. Not the interactions between them. (see mocking and patching)
```
[the remark this answers is not in the document]
Comment (henryiii) [reply docs/pages/principles/testing.md:41]: Also, I think this is easy to misinterpret. For example, if you have this simplified example:

```python
@dataclasses.dataclass
class Vector:
    x: float
    y: float

def mag2(vec: Vector) -> float:
    return vec.x**2 + vec.y**2
```

If you took this at face value, that would mean you'd need to mock a Vector class to test mag2; they are two units. I think it would be good to clarify "interactions between them", or give an example, or mention that things like the above are fine. For the above case, you'd want to test `Vector` on it's own, then it's fine to use it to test other units that use it.

**Comment by henryiii** [comment]
Here's an example I added to boost-histogram: https://github.com/scikit-hep/boost-histogram/pull/1022

**Comment by lundybernard** [comment]
The term I use is "Diagnostic Tests", 
The audience for diagnostic tests is people checking or troubleshooting live / production deployments who need to verify that an installation is correct.

FFR: 
"Smoke tests" are a sub-set of test cases, that are run to quickly verify that key features of a system work.
We some times resort to creating a smoke test suite out of desperation, when the runtime of the full test suite becomes a burden.

File: docs/pages/principles/testing.md (hunk 6 above)
Comment (lundybernard) [reply docs/pages/principles/testing.md]: Quick explanation of where this suggestion came from, and why I agree that it should not be recommended in this context:

The motivation is to test the python source in complete isolation, allowing us to pull the repo, and run tests without any dependencies... 
Its a micro-optimization for CI, which may save you from installing dependencies when an error is found in the code.

In retrospect, this is a bit silly:
* It would force excessive mocking of dependencies.
* The time & bandwidth cost of creating python environments, is less of a concern today.
* I think the added complexity out-weighs any advantages.

I suggest adding a section to the `Intro to development` section to describe the development workflow.
Something like:

1. create or clone the repository
1. create a virtual environment
1. build and install the package
1. run unit tests (and/or diagnostic tests) to verify the install
1. run the full test suite
1. make code changes
1. rebuild binaries *if needed*
1. run unit tests (to fail fast)
1. run the full test suite
1. commit and push updates
1. make coffee while CI runs

File: docs/pages/principles/testing.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (henryiii) [reply docs/pages/principles/testing.md]: That's what I meant, it's not a hard and fast rule, so it's better to explain why it's important (isolating interactions, avoiding side effects, file access, etc) so the developer can use discretion. Most data classes (the style, not necessarily from dataclasses.dataclass) don't *need* to be mocked. While classes with non-trivial inits, file access, etc. are better mocked.

File: docs/pages/principles/testing.md (hunk 6 above)
Comment (henryiii) [reply docs/pages/principles/testing.md]: I agree this needs to be expanded and worked on, but I've not been very sure where. There's https://learn.scientific-python.org/development/tutorials/dev-environment/, which does a bit of this, but needs work (and doesn't include the whole workflow). It also should show the modern workflow, which is just:

```
uv run pytest
```

Which does steps 2, 3, and 4 all at once for you (and steps 7 and 8 for you if rerun).

And uv-based CI for simple projects doesn't give you time for coffee. A full job can run in 10-15 seconds. :)

Comment (henryiii) [reply docs/pages/principles/testing.md]: Oh, and the reason I'm not sure where (since that's pretty much a perfect spot) is because there are now other nearly identical pages:

* https://learn.scientific-python.org/contributors/setup/install/
* https://github.com/scientific-python/learn.scientific-python.org/pull/247

File: docs/pages/principles/testing.md [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+failure. [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) is an
+excellent tool for this, and a lot of fun to use.
+
+- SLOW TESTS: fuzz tests can take a very long time to run, and should usually be
+  placed in a test suite, which we run separately from our faster tests.
+- Reserve fuzz testing for the few critical functions, where it really matters.
+
+#### Integration Tests
+
+The word "Integration" is a bit overloaded, and can reffer to many levels of
```
Comment (henryiii) [inline docs/pages/principles/testing.md:90]:
```suggestion
The word "Integration" is a bit overloaded, and can refer to many levels of
```

**Comment by lundybernard** [comment]
@henryiii this is ready for another review

**Comment by henryiii** [comment]
For now, keeping it on one page makes sense, as there's a discussion on restructuring going on anyway. I'm about to leave for a week-long vacation, so might not be able to get to this. I've got a couple of sections I'd like to work on a bit, and I'd like to compare this with the current pytest page to see if I can reduce overlap or cross-link.

One solution could be to merge in more-or-less the current state (I could do a quick review), then I could make PRs to update it later. Or we can wait till I get back - do you have a preference?

**Comment by lundybernard** [comment]
This went through several rounds of proof-reading and revision, so I'm happy to merge it, and continue working on it in the future through new Issues/PRs.

I don't mind if you prefer to wait and do a more thorough final review when you get back.  Whatever works best for you.

File: docs/pages/principles/testing.md [hunk 11]
```diff
@@ -0,0 +1,633 @@
...
+    assert ret == "hello world"
+    mock_dangerous_sideffects.assert_called_once()
+
+    # But this will still raise an exception!
+    dangerous_sideffects()
+```
+
+## Testing Edgecases
+
+While writing unit tests, you may be tempted to test edgecases. You may have a
```
[the remark this answers is not in the document]
Comment (lundybernard) [reply docs/pages/principles/testing.md:487]: Maybe this needs a little more elaboration.

I think excessive edge case testing in unit tests is bad, we do want to test a reasonable subset of expected, and supported, valid inputs.  This may be a good place to mention the TDD rule about not testing against invalid input until it actually becomes a problem.  That said, I would not complain if someone showed me a huge test suite for a critical function that still runs in less than a second.

I'll work on the wording in this section.

File: docs/pages/principles/testing.md [hunk 12]
```diff
@@ -0,0 +1,633 @@
...
+- Makes using drop-in replacements simpler. Changing
+  `from pandas import DataFrame` to `from polars import DataFrame` in your
+  file-under-test, should result in all tests passing, with no other changes.
+
+It is common practice to import and alias all of a library, such as
+`import numpy as np`. However, as we develop our unit tests, this can cause
+difficulty with mocking, and complicate refactoring.
+
+To patch out numpy.sum in your test, you either need to patch the _global_ numpy
+module, which can have unintended side-effects, or specifically patch numpy.sum
```
[the remark this answers is not in the document]
Comment (lundybernard) [reply docs/pages/principles/testing.md:355]: This needs some clarification.  Yes, the lifetime of the patch is limited to the lifetime of the function, but the patch can target the namespace of the file-under-test, or the namespace of the dependency.

I can provide some examples to illustrate, but they are a bit long.  This detail probably bares some detailed explanation, along with a discussion on when to use a Mock, vs when to substitute in a fake.

File: docs/pages/principles/testing.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (lundybernard) [reply docs/pages/principles/testing.md]: I rewrote this section, and changed the title to "Extensive Input Testing".
The rewrite emphasizes readability, and recommend organizing edge case and extensive input testing in a way that preserves it.

**Review by henryiii — APPROVED** [review]
Just had a quick last look, I think it's good! We can make PRs for followups as needed. Thanks for all the work on this!
